File: packages/guards/src/schema-guard.ts
import type { Guard, GuardResult, GuardContext } from '@open-guardrail/core';

interface JsonSchemaProperty {
  type: 'string' | 'number' | 'boolean' | 'object' | 'array';
}

interface SimpleJsonSchema {
  type: 'object';
  properties: Record<string, JsonSchemaProperty>;
  required?: string[];
}

interface SchemaGuardOptions {
  schema: SimpleJsonSchema;
  action: 'block' | 'warn';
}

function validateSchema(data: unknown, schema: SimpleJsonSchema): string[] {
  const errors: string[] = [];
  if (typeof data !== 'object' || data === null || Array.isArray(data)) {
    errors.push('Expected an object');
    return errors;
  }
  const obj = data as Record<string, unknown>;

  for (const key of schema.required ?? []) {
    if (!(key in obj)) errors.push(`Missing required field: ${key}`);
  }

  for (const [key, prop] of Object.entries(schema.properties)) {
    if (key in obj && typeof obj[key] !== prop.type) {
      errors.push(`Field "${key}" expected ${prop.type}, got ${typeof obj[key]}`);
    }
  }
  return errors;
}

export function schemaGuard(options: SchemaGuardOptions): Guard {
  return {
    name: 'schema',
    version: '0.1.0',
    description: 'JSON Schema validation guard',
    category: 'format',
    supportedStages: ['output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();

      let parsed: unknown;
      try {
        parsed = JSON.parse(text);
      } catch {
        return {
          guardName: 'schema', passed: false, action: options.action,
          latencyMs: Math.round(performance.now() - start),
          details: { reason: 'JSON parse error' },
        };
      }

      const errors = validateSchema(parsed, options.schema);
      const triggered = errors.length > 0;

      return {
        guardName: 'schema',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { errors } : undefined,
      };
    },
  };
}

File: python/open_guardrail/guards/schema_guard.py
"""Validate JSON against a simple schema."""

from __future__ import annotations

import json
import time
from typing import List

from open_guardrail.core import GuardResult


class _SchemaGuard:
    def __init__(
        self,
        *,
        action: str = "block",
        required_keys: List[str],
        optional_keys: List[str] | None = None,
    ) -> None:
        self.name = "schema-guard"
        self.action = action
        self._required = required_keys
        self._optional = optional_keys or []

    def check(
        self, text: str, stage: str = "output"
    ) -> GuardResult:
        start = time.perf_counter()

        try:
            data = json.loads(text.strip())
        except (json.JSONDecodeError, TypeError):
            elapsed = (
                time.perf_counter() - start
            ) * 1000
            return GuardResult(
                guard_name="schema-guard",
                passed=False,
                action=self.action,
                message="Invalid JSON",
                latency_ms=round(elapsed, 2),
            )

        if not isinstance(data, dict):
            elapsed = (
                time.perf_counter() - start
            ) * 1000
            return GuardResult(
                guard_name="schema-guard",
                passed=False,
                action=self.action,
                message="JSON root is not an object",
                latency_ms=round(elapsed, 2),
            )

        missing = [
            k
            for k in self._required
            if k not in data
        ]
        allowed = set(self._required) | set(
            self._optional
        )
        extra = [
            k for k in data if k not in allowed
        ]

        triggered = len(missing) > 0
        elapsed = (time.perf_counter() - start) * 1000
        return GuardResult(
            guard_name="schema-guard",
            passed=not triggered,
            action=(
                self.action if triggered else "allow"
            ),
            message=(
                "Missing keys: " + ", ".join(missing)
                if triggered
                else None
            ),
            latency_ms=round(elapsed, 2),
            details=(
                {
                    "missing": missing,
                    "extra": extra,
                }
                if triggered or extra
                else None
            ),
        )


def schema_guard(
    *,
    action: str = "block",
    required_keys: List[str],
    optional_keys: List[str] | None = None,
) -> _SchemaGuard:
    return _SchemaGuard(
        action=action,
        required_keys=required_keys,
        optional_keys=optional_keys or [],
    )
